db/db_impl/db_impl_open.cc

```diff
@@ -1158,266 +1171,470 @@ Status DBImpl::RecoverLogFiles(const std::vector<uint64_t>& wal_numbers,
-        assert(immutable_db_options_.wal_recovery_mode ==
-                   WALRecoveryMode::kTolerateCorruptedTailRecords ||
-               immutable_db_options_.wal_recovery_mode ==
-                   WALRecoveryMode::kAbsoluteConsistency);
+    reporter->status = nullptr;
+  } else {
+    reporter->status = reporter_status;
+  }
+  // We intentially make log::Reader do checksumming even if
+  // paranoid_checks==false so that corruptions cause entire commits
+  // to be skipped instead of propagating bad information (like overly
+  // large sequence numbers).
+  reader.reset(new log::Reader(immutable_db_options_.info_log,
+                               std::move(file_reader), reporter,
+                               true /*checksum*/, wal_number));
+  return status;
+}
+
+Status DBImpl::ProcessLogRecord(
+    Slice record, const std::unique_ptr<log::Reader>& reader,
+    const UnorderedMap<uint32_t, size_t>& running_ts_sz, uint64_t wal_number,
+    const std::string& fname, bool read_only, int job_id,
+    std::function<void()> logFileDropped, DBOpenLogReporter* reporter,
+    uint64_t* record_checksum, SequenceNumber* next_sequence,
+    bool* stop_replay_for_corruption, Status* status,
+    bool* stop_replay_by_wal_filter,
+    std::unordered_map<int, VersionEdit>* version_edits, bool* flushed) {
+  assert(reporter);
+  assert(stop_replay_for_corruption);
+  assert(status);
+  assert(stop_replay_by_wal_filter);
+
+  Status process_status;
+  bool has_valid_writes = false;
+  WriteBatch batch;
+  std::unique_ptr<WriteBatch> new_batch;
+  WriteBatch* batch_to_use = nullptr;
+
+  if (record.size() < WriteBatchInternal::kHeader) {
+    reporter->Corruption(record.size(),
+                         Status::Corruption("log record too small"));
+    assert(process_status.ok());
+    return process_status;
+  }
+
+  process_status = InitializeWriteBatchForLogRecord(
+      record, reader, running_ts_sz, &batch, new_batch, batch_to_use,
+      record_checksum);
+  if (!process_status.ok()) {
+    return process_status;
+  }
+  assert(batch_to_use);
+
+  SequenceNumber sequence = WriteBatchInternal::Sequence(batch_to_use);
+  if (sequence > kMaxSequenceNumber) {
+    reporter->Corruption(
+        record.size(),
+        Status::Corruption("sequence " + std::to_string(sequence) +
+                           " is too large"));
+    assert(process_status.ok());
+    return process_status;
+  }
+
+  if (immutable_db_options_.wal_recovery_mode ==
+      WALRecoveryMode::kPointInTimeRecovery) {
+    MaybeReviseStopReplayForCorruption(sequence, next_sequence,
+                                       stop_replay_for_corruption);
+    if (*stop_replay_for_corruption) {
+      logFileDropped();
+      assert(process_status.ok());
+      return process_status;
+    }
+  }
+
+  // For the default case of wal_filter == nullptr, always performs no-op
+  // and returns true.
+  if (!InvokeWalFilterIfNeededOnWalRecord(wal_number, fname, *reporter, *status,
```

Comment:
> seems to me it's copying a status into itself

```
bool DBImpl::InvokeWalFilterIfNeededOnWalRecord(uint64_t wal_number,
                                                const std::string& wal_fname,
                                                log::Reader::Reporter& reporter,
                                                Status& status,
                                                bool& stop_replay,
                                                WriteBatch& batch)
```
It's the & reference so it can directly change the status, not a copy. It's confusing at first you are right.....